### PR TITLE
Add action flag to eval

### PR DESCRIPTION
**Pull request by martmists**
- Implements `Literal` (requires 3.7+, implementation by Danny#0007)
- Hourly job to remove unused contexts

### CODE REVIEW

File: bot/cogs/snekbox.py [hunk 1]
```diff
@@ -196,6 +219,15 @@ def get_status_emoji(results: dict) -> str:
             await ctx.invoke(self.bot.get_command("help"), "eval")
             return
 
+        if action == "-save":
```
Comment (scragly) [inline bot/cogs/snekbox.py]: This entire thing appears to be a poor-mans subcommand handler. There's no reason to parse distinct flags contrary to the entire commands framework design when we have the ability to add subcommands through usage of a command group object.

File: bot/cogs/snekbox.py [hunk 2]
```diff
@@ -43,6 +46,25 @@ def __init__(self, bot: Bot):
         self.bot = bot
         self.jobs = {}
 
+        # Dict[int[user_id],
+        #      Tuple[str[saved_code],
+        #            datetime[last_used]]]
```
Comment (scragly) [inline bot/cogs/snekbox.py]: Don't randomly leave commented out code in your PRs.

File: bot/cogs/snekbox.py [hunk 3]
```diff
@@ -177,7 +199,8 @@ def get_status_emoji(results: dict) -> str:
     @command(name="eval", aliases=("e",))
     @guild_only()
     @in_channel(Channels.bot, hidden_channels=(Channels.esoteric,), bypass_roles=EVAL_ROLES)
-    async def eval_command(self, ctx: Context, *, code: str = None) -> None:
+    async def eval_command(self, ctx: Context, action: Optional[Literal["-save", "-load"]] = None,
+                           *, code: str = None) -> None:
```
Comment (scragly) [inline bot/cogs/snekbox.py]: This does not conform to typical project style. Multiline arguments are dropped after the brackets:
```py
    async def eval_command(
        self, ctx: Context, action: Optional[Literal["-save", "-load"]] = None, *, code: str = None
    ) -> None:
```

File: bot/converters.py [hunk 4]
```diff
@@ -259,3 +259,24 @@ class ISODateTime(Converter):
...
+        if not isinstance(item, tuple):
+            item = (item,)
+
+        class LiteralProxy(Converter):
+            @classmethod
+            async def convert(cls, ctx, argument):
+                if argument in item:
+                    return argument
+                raise BadArgument(f"Expected literal: one of {list(map(repr, self.literals))}")
+        return LiteralProxy
```
Comment (scragly) [inline bot/converters.py]: Unnecessary to add at this point in time due to suggesting to change commands processing to the existing command groups format.

File: bot/converters.py [hunk 5]
```diff
@@ -259,3 +259,24 @@ class ISODateTime(Converter):
             dt = dt.replace(tzinfo=None)
 
         return dt
+
+
+class Literal:
+    """
+    Matches a literal, useful for e.g. flags.
+    Can be used with Optional.
```
Comment (scragly) [inline bot/converters.py]: Linter: `D205 1 blank line required between summary line and description`

File: bot/cogs/snekbox.py [hunk 6]
```diff
@@ -43,6 +46,25 @@ def __init__(self, bot: Bot):
         self.bot = bot
         self.jobs = {}
 
+        # Dict[int[user_id],
+        #      Tuple[str[saved_code],
+        #            datetime[last_used]]]
+        self.contexts = {}
+        self._run = True
+
+        async def clean_task():
```
Comment (scragly) [inline bot/cogs/snekbox.py]: Linter: `TYP201 Missing return type annotation for public function`

Also where's the docstring here.

File: bot/cogs/snekbox.py (hunk 2 above)
Comment (martmists) [reply bot/cogs/snekbox.py]: This isn't commented out code, this is a comment explaining the type used so other people working on this know what it is.

File: bot/cogs/snekbox.py (hunk 3 above)
Comment (martmists) [reply bot/cogs/snekbox.py]: Doesn't this fprmat violate PEP8?

Comment (scragly) [reply bot/cogs/snekbox.py]: No.
> The closing brace/bracket/parenthesis on multi-line constructs may either line up under the first non-whitespace character of the last line of li

**Pull request closed without merging**